Membership Block

artpi opened the pull request:
## This is the first PR of the block that is intended to provide something that beta users can test. We intend to work in small iterations and we will gladly address feedback in later PRs. 

This introduces membership-button block to Jetpack.
This essentially will be a recurring donation button.

Purpose of the code in this PR is to:

- introduce endpoints that communicate with WPCOM
- whitelist certain options, CPTs and meta to store / sync data,
- introduce gutenberg block that uses these endpoints and provides UI to connect to Stripe, create and choose a product
- introduce a frontend of a block with the sole purpose of displaying a checkout window from wpcom in an iframe

~This needs D26086-code .~ - deployed


## Screenshots

![Zrzut ekranu 2019-03-27 o 18 17 26](https://user-images.githubusercontent.com/3775068/55098446-0df9e500-50be-11e9-99fb-5ffa6b6b55b9.png)


![Zrzut ekranu 2019-03-27 o 17 41 59](https://user-images.githubusercontent.com/3775068/55097068-64195900-50bb-11e9-8997-2d05bdc70dcc.png)

![Zrzut ekranu 2019-03-27 o 17 47 59](https://user-images.githubusercontent.com/3775068/55097075-6976a380-50bb-11e9-970a-853da7abf6b8.png)

## Internal references:

p89M8K-4a-p2
p1HpG7-6n6-p2

## Testing instructions - Jurassic Ninja

The problem with this is that you can only use real money and real stripe accounts. If you want sandbox payments, you need to use the ngrok flow

1. https://jurassic.ninja/create/?jetpack-beta&branch=add/memberships
2. Connect Jetpack, enable "BETA_BLOCKS" in /wp-admin/options-general.php?page=companion_settings
3. Go to editor
4. Search for "Membership button" block
5. You will get a placeholder that will say "connect to stripe", click that
6. If you correctly set up the store sandbox, you will get "you can skip this form" at the top, click that
7. You will get redirected to ugly looking page on success
8. Now go back to gutenberg, click "Re-check Connection", wait
9. Now you should have a button to add product, click it, fill the form
10. choose your new product
11. Publish the page
12 Go to frontend, click the button that says contribution
13. go with the flow

## Testing instructions - Jetpack Docker

1. Set up Jetpack with docker and ngrok.io, connnect jetpack
2. Remember to turn on beta blocks
```
define( 'GUTENBERG_DEVELOPMENT_MODE', true );
define( 'JETPACK_BETA_BLOCKS', true );
```
3. Checkout this PR
4. Build blocks by
```
yarn build-extensions
```
5. Continue from point 3. in the previous instruction

artpi committed:
Initial code from wpcom

artpi committed:
fix blog id

artpi committed:
Fix content

artpi committed:
Linter

artpi committed:
again - lint

artpi committed:
Rework into a block part 1

artpi committed:
memberships changes

artpi committed:
Memberships status endpoint

artpi committed:
endpoint enhancements

artpi committed:
Something memberships

artpi committed:
whitelist the option

artpi committed:
Thickbox is working

artpi committed:
Pass button color

artpi committed:
Endpoint changes because I have to rebase

artpi committed:
Port work from calypso

artpi committed:
Make  it work

artpi committed:
Fixes to the UI

artpi committed:
fix for price change

artpi committed:
remove legacy code

jeherve commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,291 @@
+<?php
+/**
```
I think you can remove that comment altogtether. Once it's in Fusion there will be a pre-commit warning when you make changes to the file on WordPress.com anyway.

artpi replied:
Done!

jeherve commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,291 @@
+	 *
+	 * @var string
+	 */
+	static private $version = '0.03';
```
Could we use `JETPACK__VERSION` here instead?

artpi replied:
Done!

jeherve commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,291 @@
+	 *
+	 * @return string
+	 */
+	private function format_price( $plan ) {
```
Could we re-use the apparently more fully-fledged `format_price` function what we already have for Simple Payments here, to avoid duplication?

simison replied:
Shall it be moved to more obvious shared location then, too?

artpi replied:
With recent changes, UI will always provide this data.
Since the entire currency formatting in PHP was just to provide a fallback, I provided a very generic one and removed alll Simple Payments dependenciec

jeherve commented on modules/module-extras.php:
```diff
@@ -23,6 +23,7 @@
 	'seo-tools/jetpack-seo-titles.php',
 	'seo-tools/jetpack-seo-posts.php',
 	'simple-payments/simple-payments.php',
+	'memberships/class-jetpack-memberships.php',
```
Since the tool requires a WordPress.com connection, maybe we could move it with the SEO tools a bit downer in that file?

artpi replied:
I think it's working in the opposite way.
The modules right here are the ones that work only if JP is active,

The seo tools down there are working only if JP is INactive:

```
// Not every tool needs to be included if Jetpack is inactive and not in development mode
if ( ! Jetpack::is_active() && ! Jetpack::is_development_mode() ) {
	$tools = array(
		'seo-tools/jetpack-seo-utils.php',
		'seo-tools/jetpack-seo-titles.php',
		'seo-tools/jetpack-seo-posts.php',
	);
}
```

jeherve replied:
Oh yeah you're right, that's confusing. Right now the default list is loaded as soon as the plugin is active, whether it's connected to WordPress.com or not, and includes things that need a connection as well as things that don't. It's not ideal. We'll need to refactor that in the future I think, to better separate the 2 scenarios. That can be done in a different PR though. 👍 	

Until then, I think it may be good to short-circuit early in `memberships/class-jetpack-memberships.php` if Jetpack is not connected to WordPress.com.

simison replied:
Should that clarification be code-commented above the list here?

jeherve replied:
I've just pushed #11796 to start reorganizing things a bit.

jeherve reviewed (changes requested):
Just a few minor comments, I have not had the time to fully test this yet.

Also, and since we are introducing a bunch a new files, what do you think about adding them all to the pre-commit hook here in Jetpack?
https://github.com/Automattic/jetpack/blob/7b6c7495020716d3d89fd7b0b8d0f4f3093348bb/bin/pre-commit-hook.js#L37

We've been trying to do that for all new files to slowly move towards better code coverage in the plugin.

jeherve commented on _inc/lib/core-api/wpcom-endpoints/memberships.php:
```diff
@@ -0,0 +1,137 @@
+/**
+ * Memberships: API to communicate with "product" database.
+ *
+ * @since 7.2
```
Since we're already past code freeze for 7.2, you'll most likely want to update references to 7.3 everywhere in that PR.

artpi replied:
Done!

jeherve commented on sync/class.jetpack-sync-defaults.php:
```diff
@@ -74,6 +74,7 @@ class Jetpack_Sync_Defaults {
 		'disabled_reblogs',
 		'jetpack_comment_likes_enabled',
 		'twitter_via',
+		'jetpack-memberships-connected-account-id',
```
Since you've added a new option here, you'll want to update the array in the tests as well to avoid failures with the unit tests:
https://github.com/Automattic/jetpack/blob/e49502abdfae4340d06ca6cd4f1662b9590978c6/tests/php/sync/test_class.jetpack-sync-options.php#L203

artpi replied:
Done!

artpi commented on extensions/blocks/membership-button/index.js:
```diff
@@ -0,0 +1,79 @@
+	save: () => null,
+};
+
+export const SUPPORTED_CURRENCY_LIST = [
```
We don't want that.
Currency list is dependant on payment processor supported currency list and it's different for simple payments and memberships.
Every one of these modules may have a different logic, because it has a different backend.
Once we do something about unifying those, then these lists will be in sync, but for now they are completely seperate.

simison replied:
>Currency list is dependant on payment processor supported currency list and it's different for simple payments and memberships.

If that's the case, you shouldn't use the default currency from simple payments either:

https://github.com/Automattic/jetpack/pull/9802/files#diff-20d5519b6b59be680f956bbd53c2f1d1R12

artpi replied:
Changed it to proper one and added a comment where this comes from.

artpi committed:
Review feedbac

davemart-in reviewed (dismissed):
Keep rocking!  🤘 From what I can tell, this is shaping up nicely.

artpi committed:
repurpose format_price

artpi committed:
permissions check

artpi committed:
remove simple payments import

artpi committed:
make it pull a proper file

jeherve commented on extensions/blocks/membership-button/membership-button.php:
```diff
@@ -7,7 +7,11 @@
+if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+	require_once WP_CONTENT_DIR . '/mu-plugins/memberships/class-jetpack-memberships.php';
+} elseif ( Jetpack::is_active() ) {
+	require_once JETPACK__PLUGIN_DIR . '/modules/memberships/class-jetpack-memberships.php';
```
Would it be an option to place the file there on wpcom as well?

artpi replied:
I did not realize we do that, I assumed all JP modules have to be in the mu-plugins.
I'll happily move it!

artpi replied:
Moved! 🚀

simison commented on extensions/blocks/membership-button/edit.jsx:
```diff
@@ -0,0 +1,337 @@
+import classnames from 'classnames';
+import SubmitButton from '../../utils/submit-button';
+import apiFetch from '@wordpress/api-fetch';
+import { __ } from '../../utils/i18n';
```
Look for https://github.com/Automattic/jetpack/pull/11743 getting merged soon; these i18n functions will need rebasing.

simison commented on extensions/blocks/membership-button/edit.jsx:
```diff
@@ -0,0 +1,337 @@
+ * Internal dependencies
+ */
+import { icon, SUPPORTED_CURRENCY_LIST } from '.';
+import { formatPrice } from '../simple-payments/utils';
```
For this re-use to be more obvious and cleaner, let's move the `formatPrice` to [shared utils folder](https://github.com/Automattic/jetpack/tree/858b174e510ebcb59904cbeb0a89a072870d33e1/extensions/utils) instead of re-using it from another block's sources.

artpi replied:
Done!

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+			$data['id'],
+			implode( $classes, ' ' ),
+			$button_styles,
+			$data['button_label']
```
Should we escape all these attributes?

sirreal replied:
❗️Yes. This is a blocker, attributes and HTML need to be properly escaped here.

artpi replied:
Done!

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+		if ( $plan['formatted_price'] ) {
+			return $plan['formatted_price'];
+		}
+		return Jetpack_Simple_Payments::format_price( $plan['price'],$plan['currency'] );
```
Re a method from another module feels little odd. @jeherve @kraftbj what's a good place for utility functions in Jetpack?

simison replied:
Note that modules have "a contract" in Jetpack where they might be turned on or off and the code might not therefore exist. That said, I think Simple payments currently is _always_ loaded, but I wouldn't take that granted in terms of code-maintenance.

jeherve replied:
We have `functions.global.php` for those things, for example. That file is not synced with WordPress.com right now, but we could copy it over.

simison replied:
Nice, that would be good move. Like this it's quite prone to break unintentionally unless it's covered with tests.

artpi replied:
👍

artpi replied:
I removed the dependency on simple payments entirely.
This was just for providing a fallback in the PHP, but with recent UI changes it is very unlikely that it will be needed.
I ended uf providing a very generic fallback.

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+			return get_current_blog_id();
+		}
+
+		return Jetpack_Options::get_option( 'id' );
```
Not a change for this PR, but just noting that feels weird that folks need to do `IS_WPCOM` for a simple thing like this — could the wpcom check be in `get_option` method directly? @jeherve

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+		$data = array(
+			'blog_id' => $this->get_blog_id(),
+			'id'	  => $id,
+			'button_label' => sprintf( __( '$%s Contribution', 'jetpack' ),$this->format_price( $plan ) ),
```
nit; space after `,`

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+	 * @return string|void
+	 */
+	public function render_button( $attrs ) {
+		Jetpack_Gutenberg::load_assets_as_required( self::$button_block_name);
```
nit; space before `)`

I think adding this new class to [linter](https://github.com/Automattic/jetpack/blob/858b174e510ebcb59904cbeb0a89a072870d33e1/bin/pre-commit-hook.js#L31) would be excellent.

artpi replied:
Everything is linted now!

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+		if ( ! $product || is_wp_error( $product ) ) {
+			return;
+		}
+		if ( $product->post_type !== self::$post_type_plan || 'trash' === $product->post_status ) {
```
Does `trash ===` cover everything security wise? Shouldn't we check for `public !==`?

artpi replied:
There is no way of making this a draft or otherwise.
Not even security is our concern, since this serves as a local cache for BD data.
Manipulating this is proxied by access methods on wpcom.
I may need to have to rework this, but definitely there is no danger here.

simison replied:
Is checking for `public !==` somehow bad idea, even if there are no security concerns?

artpi replied:
Thit has been tested multiple times and I am certain it works as intended.
I am afraid of touching it, since the post status is a very hacky thing - for example the `future` status seems like it can introduce problems in this scenario.
I would remove 'trash' check even, but this will be needed for omitting junk data created by inconsistent versions of this code.

simison replied:
I was pointing out this just because I remember we had to fix some bugs around Simple payments exactly because of similar "trash" status check. While it probably worked fine initially, original author probably hadn't seen how someone could extend the code/API and leave easy way for regressions.

That said, this case might be different and it's up to you of course. 🙂

simison replied:
(For a reference https://github.com/Automattic/wp-calypso/pull/28815/files#r235933564)

artpi replied:
Sorry, I did not follow this particular regression, but from what I understand from a quick scan:

- Data was stored in CPTs for shortcode
- Upon introducing a block, the initial solution attempted to store a copy of the same data in block attributes leading to source of truth problems, especially in block duplication
- This was resolved by introducing a "draft" status
- Which in turn, was clashing with that !== `trash` status check

Is there more to it? I'm having trouble finding a more comprehensive writeup.
I am all for avoiding regressions, but I don't want to introduce new ones while at it.

If there is an additional benefit of keeping the logic similar to simple payments without visible downsides, I'll implement it 👍

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+	 * @return array
+	 */
+	private static function get_plan_property_mapping() {
+		$meta_prefix = 'mem_';
```
Would `memberships_` be more clear and meaningful? Now folks need to look this code up to understand what this is when they see the option in database. Do we need to prefix these with `jetpack_`?

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+			'delete_post'        => 'delete_posts',
+			'edit_posts'         => 'edit_posts',
+			'edit_others_posts'  => 'edit_others_posts',
+			'publish_posts'      => 'publish_posts',
```
Does this capability make it so that the block doesn't work for contributor users? If so, does the editor reflect that graciously without failure?

artpi replied:
Since these are changed by custom API and access methods, I opted out for restricting user capabilities to change them.
It all graciously handles these cases.

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+		if ( $plan['formatted_price'] ) {
+			return $plan['formatted_price'];
+		}
+		return Jetpack_Simple_Payments::format_price( $plan['price'],$plan['currency'] );
```
Note also spacing after `,`

artpi replied:
This is outdated since I removed simple payments dependency entirely

simison commented on extensions/blocks/membership-button/view.scss:
```diff
@@ -0,0 +1,50 @@
+.jetpack-memberships-modal .TB_footer {
+	border-top: 1px solid #ddd;
+	color: #a5b2bd;
+	font-family: 'Lato', sans-serif;
```
Introducing a new font family in Jetpack/wp-dash/Gutenberg seems off, what's the idea here?

artpi replied:
Yeah, seems off to me as well, removing

simison commented on extensions/blocks/membership-button/view.scss:
```diff
@@ -0,0 +1,50 @@
+}
+.jetpack-memberships-modal .TB_footer {
+	border-top: 1px solid #ddd;
+	color: #a5b2bd;
```
We shouldn't use fixed colors like this. Please use either Gutenberg colour variables (by adding them on top of this file so that we can easily migrate them later to use Gutenberg sass files, [see example](https://github.com/Automattic/jetpack/blob/858b174e510ebcb59904cbeb0a89a072870d33e1/extensions/blocks/markdown/editor.scss#L1-L3)) or by using Muriel colour CSS variables listed here: https://github.com/Automattic/color-studio/blob/546e4ed57ec2426246049096fbd3064e1d4ecfb0/dist/color-properties.css

artpi replied:
@lessbloat any help here? :)

simison commented on extensions/blocks/membership-button/view.scss:
```diff
@@ -0,0 +1,50 @@
+#TB_window.jetpack-memberships-modal {
+	border-radius: 4px;
+	background-color: #f9f9f9;
+	background-image: url( 'https://s0.wp.com/i/loading/loading-64.gif' );
```
Any change we could use Gutenberg [Spinner](https://github.com/WordPress/gutenberg/tree/6c9a9b22312acde7e64462f554fe3b9006becc9f/packages/components/src/spinner) component instead?

artpi replied:
a) That would be impossible with thickbox
b) This is a trick, because this is really under a transparent iframe. Using anything that is not a gif or css-only background would require sorcery with z-indexes.

simison commented on extensions/blocks/membership-button/view.js:
```diff
@@ -0,0 +1,57 @@
+	mailchimpBlocks.forEach( block => {
+		const blog_id = block.getAttribute( 'data-blog-id' );
+		const plan_id = block.getAttribute( 'data-plan-id' );
+		const powered_text = block
```
Please camelCase in JS. :-)

simison commented on extensions/blocks/membership-button/view.js:
```diff
@@ -0,0 +1,57 @@
+}
+
+const initializeMembershipButtonBlocks = () => {
+	const mailchimpBlocks = Array.from( document.querySelectorAll( '.' + blockClassName ) );
```
mailchimp?

artpi replied:
Done!

simison commented on extensions/blocks/membership-button/view.js:
```diff
@@ -0,0 +1,57 @@
+
+function activateSubscription( block, blogId, planId, poweredText ) {
+	block.addEventListener( 'click', () => {
+		tb_show(
```
Could we use the [modal shipped in Gutenberg](https://github.com/WordPress/gutenberg/tree/6c9a9b22312acde7e64462f554fe3b9006becc9f/packages/components/src/modal)? Using these ancient jQuery libraries feels like we'll just introduce some serious maintenance dept.

artpi replied:
This modal is used on site frontend, not the administration part.
I opted out for not including the entirety of React library for every site visit.

simison replied:
I agree with the sentiment — did you check if enqueuing entirety of jQuery, Dashicons and Thickbox is much different tho? 🙃

simison commented on extensions/blocks/membership-button/membership-button.php:
```diff
@@ -0,0 +1,23 @@
+	require_once JETPACK__PLUGIN_DIR . '/modules/memberships/class-jetpack-memberships.php';
+}
+
+if ( ( defined( 'IS_WPCOM' ) && IS_WPCOM ) || Jetpack::is_active() ) {
```
You shouldn't need to do this check since `is_active` is done before blocks registered with `jetpack_register_block`  are loaded:

https://github.com/Automattic/jetpack/blob/858b174e510ebcb59904cbeb0a89a072870d33e1/class.jetpack-gutenberg.php#L417-L430

simison commented on extensions/blocks/membership-button/index.js:
```diff
@@ -0,0 +1,79 @@
+	save: () => null,
+};
+
+export const SUPPORTED_CURRENCY_LIST = [
```
Would be useful to indicate with code comment where this list came from and how to keep it up to date — consider future maintainers who might have no idea.

simison commented on extensions/blocks/membership-button/index.js:
```diff
@@ -0,0 +1,79 @@
+import { __, _x } from '../../utils/i18n';
+import edit from './edit';
+import './editor.scss';
+import { DEFAULT_CURRENCY } from '../simple-payments/constants';
```
Note https://github.com/Automattic/jetpack/pull/9802/files#r271168014

artpi replied:
Removed

simison commented on extensions/blocks/membership-button/index.js:
```diff
@@ -0,0 +1,79 @@
+	keywords: [
+		_x( 'sell', 'block search term' ),
+		_x( 'subscription', 'block search term' ),
+		_x( 'stripe', 'block search term' ),
```
We can leave brand name untranslated.

simison commented on extensions/blocks/membership-button/index.js:
```diff
@@ -0,0 +1,79 @@
+		},
+		submitButtonText: {
+			type: 'string',
+			default: __( 'Your donation' ),
```
Please double check that the block doesn't invalidate if leaving this string as a default and then changing the site's locale between page-loads.

This string might not be translated yet in Jetpack so you'll prolly need to test with something like "Settings" or "Monday".

More pafL3P-kL-p2

sirreal replied:
I suspect this is currently safe because the default values are not serialized at the moment, however I'd still discourage this and instead handle missing values appropriately in the editor and the rendered view.

```suggestion
```

simison commented on extensions/blocks/membership-button/view.js:
```diff
@@ -0,0 +1,57 @@
+}
+
+const initializeMembershipButtonBlocks = () => {
+	const mailchimpBlocks = Array.from( document.querySelectorAll( '.' + blockClassName ) );
```
Make sure to add `wp-polyfill` as a dependency because of `Array.from` which doesn't work in IE11.

artpi replied:
I'm going to change it to `Array.prototype.slice.call` (https://davidwalsh.name/nodelist-array) since that sounds better than introducing another dependency :)

simison replied:
Sure, either way works for now. We (Gutenpack) will soonish enfoce `wp-polyfill` dependency on all blocks in Jetpack just because currently this is too human-error prone since the build process doesn't polyfill like developers might expect.

sirreal commented on extensions/blocks/membership-button/index.js:
```diff
@@ -0,0 +1,79 @@
+};
+
+export const SUPPORTED_CURRENCY_LIST = [
+	DEFAULT_CURRENCY,
```
It seems strange to depend on a Simple payments constant as part of this set of currencies. I wouldn't expect to change this block when modifying simple payments constants.

sirreal commented on extensions/blocks/membership-button/view.js:
```diff
@@ -0,0 +1,57 @@
+			activateSubscription( block, blog_id, plan_id, powered_text );
+		} catch ( err ) {
+			// eslint-disable-next-line no-console
+			console.error( 'Problem activating Membership Button ' + plan_id, err );
```
Typically we'd avoid this type of logging in production:

```suggestion
			if ( process.env.NODE_ENV !== 'production' ) {
				console.error( 'Problem activating Membership Button ' + plan_id, err );
			}
```

Is this an expected or unexpected error? Should we look at logging to a centralized place so we can monitor?

artpi replied:
But if something goes wrong, how is site owner supposed to know that anything is not as expected?

artpi commented on _inc/lib/core-api/wpcom-endpoints/memberships.php:
```diff
@@ -0,0 +1,137 @@
+			require_lib( 'memberships' );
+			$connected_destination_account_id = Jetpack_Memberships::get_connected_account_id();
+			if ( ! $connected_destination_account_id ) {
+				return new WP_Error( 'no-destination-account', __( 'Please set up Stripe account for this site first' ) );
```
Done!

sirreal commented on extensions/blocks/membership-button/view.js:
```diff
@@ -0,0 +1,57 @@
+	} );
+};
+
+if ( typeof window !== 'undefined' && typeof document !== 'undefined' ) {
```
We're seeing this over and over, maybe it's time to write a shared function (not requesting the change in this PR):

```js
function initView( blockInitCallback ) {
	if ( typeof window !== 'undefined' && typeof document !== 'undefined' ) {
		// `DOMContentLoaded` may fire before the script has a chance to run
		if ( document.readyState === 'loading' ) {
			document.addEventListener( 'DOMContentLoaded', blockInitCallback );
		} else {
			blockInitCallback();
		}
	}
}
```

artpi replied:
That would be helpful!

sirreal commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+	 *
+	 * @var string
+	 */
+	static private $version = JETPACK__VERSION;
```
Is there value in this? Is it used?
```suggestion
```

simison commented on extensions/blocks/membership-button/edit.jsx:
```diff
@@ -0,0 +1,337 @@
+				>
+					{ formatPrice( parseFloat( product.price ), product.currency ) }
+				</Button>
+			) ) }{' '}
```
Why the empty `' '`? Maybe this should be left for CSS?

sirreal reviewed (commented):
I'm concerned about pulling in functionality directly from Simple Payments.

sirreal commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,283 @@
+	 * @return string|void
+	 */
+	public function render_button( $attrs ) {
+		Jetpack_Gutenberg::load_assets_as_required( self::$button_block_name);
```
Missing deps?
```suggestion
		Jetpack_Gutenberg::load_assets_as_required( self::$button_block_name, array( 'thickbox', 'wp-polyfill' ) );
```

artpi replied:
Done, thank you!

sirreal commented on extensions/blocks/membership-button/view.js:
```diff
@@ -0,0 +1,57 @@
+/* global tb_show */
```
Where does this come from? Does `thixbox` need to be a view script dependency?

artpi replied:
Yes, because it is responsible for displaying an iframe with checkout flow to the actual site visitor

artpi committed:
review feedback - JS parts

sirreal committed:
Update modules/memberships/class-jetpack-memberships.php

Co-Authored-By: artpi <artur.piszek@gmail.com>

sirreal committed:
Update extensions/blocks/membership-button/edit.jsx

Co-Authored-By: artpi <artur.piszek@gmail.com>

sirreal committed:
Update extensions/blocks/membership-button/edit.jsx

Co-Authored-By: artpi <artur.piszek@gmail.com>

artpi committed:
textdomain

artpi committed:
remove unnecessary version

sirreal committed:
Remove default button text from block default

Co-Authored-By: artpi <artur.piszek@gmail.com>

sirreal committed:
Use unwrapped i18n depenencies instaed of relative path

Co-Authored-By: artpi <artur.piszek@gmail.com>

sirreal committed:
Textdomain

Co-Authored-By: artpi <artur.piszek@gmail.com>

artpi committed:
texdomain again

artpi committed:
properly escape attributes

artpi committed:
reuse currency functins

artpi committed:
Revert simple payments format_currency changes

artpi committed:
solve cancel and a small UI issue

artpi committed:
fix lint in class memberships

artpi committed:
lint api files

artpi committed:
Add php files to linter

artpi committed:
jetpack-memberships class moved on wpcom

artpi committed:
make call to wpcom as a user

artpi committed:
now really make it as a user, this is the php file

artpi commented:
> Also, and since we are introducing a bunch a new files, what do you think about adding them all to the pre-commit hook here in Jetpack?

Linted everything and added files to linters! :)

artpi committed:
Add missing textdomains

artpi committed:
Handle message to pass the iframe modal

Commit:
Update color hex's to become Muriel variables

@simison I did not see where
https://github.com/Automattic/color-studio/blob/546e4ed57ec2426246049096
fbd3064e1d4ecfb0/dist/color-properties.css was being pulled in. I sort
of did a half and half approach to what you suggested. Happy to change
this up based on your feedback.

artpi committed:
Review feedback - some protection

artpi committed:
Rename meta name from mem_ to jetpack_memberships_

simison commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,289 @@
+	 * @return string|void
+	 */
+	public function render_button( $attrs ) {
+		Jetpack_Gutenberg::load_assets_as_required( self::$button_block_name, array( 'thickbox', 'wp-polyfill' ) );
```
Note that this function suffers from regression which won't allow `thickbox` to be enqueued, I pointed it out here:

https://github.com/Automattic/jetpack/pull/11591/files#r272845122

We'll need to get that fixed for this PR.

`wp-polyfill` dependency can now be removed; we're including it with every block much like Gutenberg does.

sirreal replied:
In the near-term, I'd like to extend the webpack plugin's API to handle more scripts: https://github.com/Automattic/wp-calypso/issues/32097

For this PR, we should restore the second parameter so we can pass through the `thickbox` dependency 👍 

That will allow this PR to proceed and not rush the webpack plugin API change.

sirreal replied:
https://github.com/Automattic/jetpack/pull/11968

simison commented on extensions/blocks/membership-button/view.scss:
```diff
@@ -0,0 +1,55 @@
+}
+#TB_window.jetpack-memberships-modal {
+	border-radius: 4px;
+	background-color: $--muriel-gray-0;
```
Seems like in Jetpack these kind of SASS variables for Muriel colours work nowadays without needing to define them on top of the file: `$muriel-gray-0`.

(CSS variables like `--muriel-gray-0` used to work but looks like that's no more the case)

simison replied:
Oh seems like `var( --muriel-gray-0 )` might work. We'll have to document this stuff better. :-)

eoigal commented:
From a billing point of view, I approve this (pending changes) 👍

artpi committed:
change the var

artpi committed:
change CSS vars

artpi committed:
Fixes for changes in the upstream

artpi committed:
fix after implementing review feedback

dereksmart commented:
@artpi I noticed you deployed the wpcom side of this today in r190724-wpcom 

Is this ready to go too?

artpi commented:
@dereksmart by all means, yes!
- Deployed accompanying diff
- This is rebased and tested
- updated testing instructions
- @eoigal gave his blessing https://github.com/Automattic/jetpack/pull/9802#issuecomment-482667060
- implemented every single feedback point

kraftbj commented:
> You will get redirected to ugly looking page on success.
which is 
> You have created account #7. This will be a redirect, but now I have no place to redirect you to.[]

What's the plan for this? I'd want something nicer before merging this.

![2019-04-24 at 12 23](https://user-images.githubusercontent.com/88897/56679978-de71d480-668b-11e9-826e-cef0bacd8ba4.jpg)

"Describe your item" feels weird. It's a membership? but what's the item?

I stumbled a bit by not choosing a dollar amount and confused when nothing was rendered. Not sure if there's a way around that. I guess I expected the block to provide a dropdown of amounts (thinking of it like a donation/tip button, not just preset membership amounts).

Is there a way to remove a dollar amount? 
![2019-04-24 at 12 36](https://user-images.githubusercontent.com/88897/56680834-a7042780-668d-11e9-8974-d825f8b02423.jpg)

If I messed up when adding the $1 item, any way to remove it?

artpi commented:
Thanks @kraftbj !
All these concerns are being addressed elsewhere.
> What's the plan for this? I'd want something nicer before merging this.

This is being addressed in D26489-code which is ready and only needs 👀 .

> "Describe your item" feels weird. It's a membership? but what's the item?

@lessbloat -  any idea?

> Is there a way to remove a dollar amount?

There is a whole Management UI being built in calypso, since this seems like a complicated task to carry out within a block.
Here is the PR that introduces the section. Product management UI will follow once this gets in:

https://github.com/Automattic/wp-calypso/pull/32145

davemart-in commented:
> "Describe your item" feels weird. It's a membership? but what's the item?

Perhaps "subscription" instead of "item"?

We could probably also change the placeholder text to be "Subscription description" instead of duplicating the label above it. 👍

artpi committed:
Change copy for item description

artpi commented:
Thanks Dave!

> Perhaps "subscription" instead of "item"?
> We could probably also change the placeholder text to be "Subscription description" instead of duplicating the label above it. 👍

Both implemented in https://github.com/Automattic/jetpack/pull/9802/commits/39301fd5fcdf2d003f31f1ff48f8b0d04ed6d3a8

jeherve commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,289 @@
+			'blog_id'      => self::get_blog_id(),
+			'id'           => $id,
+			'button_label' => __( 'Your contribution', 'jetpack' ),
+			'powered_text' => __( 'Powered by WordPress.com', 'jetpack' ),
```
Should it say "Powered by Jetpack" for Jetpack sites?

artpi replied:
No, it should not. This is for the purpose of alerting subscribers that they will need wordpress.com account to facilitate the payment.

jeherve commented on extensions/blocks/membership-button/view.js:
```diff
@@ -0,0 +1,76 @@
+	block.addEventListener( 'click', () => {
+		tb_show(
+			null,
+			'https://subscribe.wordpress.com/memberships/?blog=' +
```
If my site is in French (or any other language supported by Stripe), could we pass an extra parameter to get a Stripe interface in French?

artpi replied:
Will do!

artpi replied:
Done!

jeherve commented on _inc/lib/core-api/wpcom-endpoints/memberships.php:
```diff
@@ -0,0 +1,170 @@
+			return $product->to_array();
+		} else {
+			$blog_id  = Jetpack_Options::get_option( 'id' );
+			$response = Jetpack_Client::wpcom_json_api_request_as_user(
```
This causes issues when someone who did not connect their account to WordPress.com tries to add a button. They receive the generic error below because their account is not connected to WordPress.com (even if the site is actually connected).

artpi replied:
Good catch!
FWIW I tried using `wpcom_json_api_request_as_blog` and it is broken - does not work with v2 endpoints. But for now, I'll think of how we can limit access to this or provide a more meaningful error.
The multiple user scenario on JP needs some case studies and there competing theories on how this should work.

artpi replied:
Added better feedback.
https://github.com/Automattic/jetpack/pull/9802/commits/856fd58a6c0655173e86abc1033eed8eea5682f8

jeherve reviewed (dismissed):
I could not fully test this since Stripe is not available in my country, but I left a few comments below.

***********

![image](https://user-images.githubusercontent.com/426388/56686346-1270e800-66d4-11e9-95f3-b121ac683928.png)

This talks about Memberships, and then about Stripe. What is Stripe? I am not sure I really understand what this block is about on a first glance. The block description does not tell me much more:

> Button allowing you to sell subscription products.

Could we expand on this a bit?

******

When I first created an account on Stripe, I got redirected to `public-api.wordpress.com` where all I could see was a blank page with an empty array `[]`.

I tried again, but since I did not want to create yet another new account, I tried to log in instead. I was not able to log in right away, so chose to head to my inbox, where I had an email inviting me to confirm my email address. This seemed to help.

*********

Sometimes we use `Membership`, sometimes `Memberships`. 

![image](https://user-images.githubusercontent.com/426388/56688687-9d081600-66d9-11e9-8663-8190c3b4dbe9.png)

***********

What does the "Products" sidebar do? Is a Product a Membership?

![image](https://user-images.githubusercontent.com/426388/56688746-c1fc8900-66d9-11e9-9332-380e16eddde7.png)

After I saved my first item, that menu started to make more sense I think.

<img width="279" alt="screenshot 2019-04-24 at 21 46 29" src="https://user-images.githubusercontent.com/426388/56689084-8615f380-66da-11e9-9d78-3860fcd30a55.png">

Are "subscription products", "products", and "Membership plan" the same thing?

************

I provided a price in Euros, but when I saved it was displayed in dollars:

![image](https://user-images.githubusercontent.com/426388/56688821-f4a68180-66d9-11e9-8b06-d1552795953b.png)

*************

The currency format seems wrong to me. I don't believe the Euro symbol should appear before the amount here:

![image](https://user-images.githubusercontent.com/426388/56689599-c3c74c00-66db-11e9-84c3-8428f256fcdc.png)

You can compare with the Simple Payments button for example:

![image](https://user-images.githubusercontent.com/426388/56689918-726b8c80-66dc-11e9-83c2-81e11eda8654.png)

*************

Since we have information about the user's connection to Stripe and their Stripe account, could we change the default currency to match their account's?

*************

I provided descriptions for my "Memberships Amounts" (is it the same thing as a "Product"?), but I don't seem to find those descriptions anywhere afterwards? It would be nice if they appeared next to the listings, to help me recognize which is which. It would also help if the renewal interval was displayed as well.

*************

If I click to add a new product within a block where I already added a product, the description is carried over:

![image](https://user-images.githubusercontent.com/426388/56689373-38e65180-66db-11e9-9632-19e2a39adfa2.png)

*************

Whenever I create a new product, it triggers the following errors:

```
PHP Notice:  Undefined property: stdClass::$meta_key in json-endpoints/class.wpcom-json-api-update-post-v1-1-endpoint.php on line 815
PHP Notice:  Undefined property: stdClass::$meta_key in json-endpoints/class.wpcom-json-api-update-post-v1-1-endpoint.php on line 816
PHP Notice:  Undefined variable: post in json-endpoints/class.wpcom-json-api-update-post-v1-1-endpoint.php on line 881
PHP Notice:  Trying to get property 'post_status' of non-object in json-endpoints/class.wpcom-json-api-update-post-v1-1-endpoint.php on line 881
```

*************

Where do I go to check the status of my site's connection to Stripe? If I want to disconnect, where can I do that?

This can be problematic on sites with multiple authors for example. The first author to create the Stripe connection will be getting the payments, and no other user can take that away from them or find out which user is collecting the payments.

*************

All in all, I find it a bit odd that this is a block available on the Free plan, when the Simple Payments block requires a Premium plan. They provide similar services, it feels like they should be on the same plan level.

simison commented on extensions/blocks/membership-button/membership-button.php:
```diff
@@ -0,0 +1,17 @@
+
+require_once JETPACK__PLUGIN_DIR . '/modules/memberships/class-jetpack-memberships.php';
+
+jetpack_register_block(
```
I think the generic `is_active` before _any_ blocks are loaded should protect from this?

https://github.com/Automattic/jetpack/blob/e54806a976d63db1c66f295187b2f507ba2e449e/class.jetpack-gutenberg.php#L401

jeherve replied:
No, because in this case we don't want the block to load when we are in development mode (but for some of our other blocks, like Business blocks, we do).

artpi replied:
Done!

artpi commented:
@jeherve , thanks for feedback! These are good comments.
The tricky part of this, is that no codebase has a full overview of the work being done, as this project spans billing systems, jetpack, wpcom, checkout form and the block.

What we are trying to achieve in this PR is to get the block to early beta WPCOM users, so we can get these comments from them. We are aware of all issues regarding copy, phrasing, etc. But we are pursuing a select group of users to test it first.
Unfortunately, the process is laid in a way that we cannot do so without merging this block into Jetpack, as a beta block.
So we are 100% aware of the issues you are pointing out. However, instead of giving it our best guess we'd like to iron out those kinks with the users.
We hoped to have this design issues out of the way, by putting the design discussion up front in p1HpG7-6n6-p2

> This talks about Memberships, and then about Stripe. What is Stripe? I am not sure I really understand what this block is about on a first glance. The block description does not tell me much more:

Our previous solution to that problem was to implement setup in a different place - namely the earn section, where that would be in context.
Yes. It could be explained later and we plan to do so in the following iterations. If @lessbloat can whip up a better copy now, Ill happily implement that.

> When I first created an account on Stripe, I got redirected to public-api.wordpress.com where all I could see was a blank page with an empty array [].

Yes, @kraftbj found this as well -  this is solved in D26489-code , which has not yet landed


> Since we have information about the user's connection to Stripe and their Stripe account, could we change the default currency to match their account's?

Very nice idea for the next version! Since this will be tricky, it is not our priority though.
I made a trello card: https://trello.com/c/mzYNxul6

> I provided descriptions for my "Memberships Amounts" (is it the same thing as a "Product"?), but I don't seem to find those descriptions anywhere afterwards? It would be nice if they appeared next to the listings, to help me recognize which is which. It would also help if the renewal interval was displayed as well.

There is a full-blown management UI being developed in Automattic/wp-calypso#32145 .
For V1 we introduce only a quick setup in the block. Full-blown UI will be in calypso.

However, introducing only amounts is a conscious design decision by @lessbloat 


> If I click to add a new product within a block where I already added a product, the description is carried over:

Nice catch! Will fix.

> Whenever I create a new product, it triggers the following errors:

Interesting, haven't seen it before. Good catch!

> Where do I go to check the status of my site's connection to Stripe? If I want to disconnect, where can I do that?

Oh definitely! And we gave it considerable thought - as you mentioned - multi-user sites will be tricky.

This will addressed in Earn / Memberships, which is being introduced in Automattic/wp-calypso#32145 

> All in all, I find it a bit odd that this is a block available on the Free plan when the Simple Payments block requires a Premium plan. They provide similar services, it feels like they should be on the same plan level.

It's important to note that this PR introduces nothing. Block is still in beta, but this is needed to lay down the further groundwork. No user will have access to it for now.
Secondly - the question of a free plan is being decided in https://trello.com/c/BaOYUR8x

This has been discussed with the broader marketing team. Since Dotcomission is moving towards pricing that will align our success with customer's success, this will be one of the first products that has built in tiered revenue sharing scheme:

- Blogger: 10% commission
- Personal: 8% commission
- Premium: 4% commission
- Business: 2% commission
- E-commerce: no commission

Decision what to do with simple payments has not yet been reached - this was postponed until Memberships is out in the wild and we have some data.

artpi commented on bin/phpcs-whitelist.js:
```diff
@@ -13,4 +13,7 @@ module.exports = [
 	'modules/theme-tools/social-menu/',
+	'modules/memberships/class-jetpack-memberships.php',
+	'_inc/lib/core-api/wpcom-endpoints/memberships.php',
+	'extensions/blocks/membership-button/membership-button.php',
```
Will do!

artpi replied:
Sorted!

artpi commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -0,0 +1,289 @@
+	 * @param WP_Post $product_post - CPT representing the product.
+	 * @return array
+	 */
+	public static function product_post_to_array( $product_post ) {
```
This file is verbatim in wpcom codebase and it's used there.

jeherve replied:
I seem to see 2 copies on WordPress.com, one using that function and the other that does not. Is that on purpose?

artpi replied:
Yes. One is old (mu-plugins/memberships) and with multiple PRs I am removing all usages so I can finally delete it :)

jeherve replied:
Alright. In this case I guess we'll need to either update the newest copy so it includes the shortcode (that uses the `product_post_to_array` function above) and update this PR accordingly, or remove `product_post_to_array` since it does not appear to be in use in the newer version?

artpi replied:
You are right, we should be able to safely remove this entirely. Will do!

artpi replied:
Function removed. https://github.com/Automattic/jetpack/pull/9802/commits/e018d63b46e84506af64ce0590373fd40745a670

jeherve committed:
Code review suggested edits - mostly copy

Co-Authored-By: artpi <artur.piszek@gmail.com>

davemart-in commented:
> However, introducing only amounts is a conscious design decision by @lessbloat

@artpi one possible workaround here might be to just toss the description in as the title attribute for each button. That way the UI stays clean, but there is additional context if you hover over a button and wait to see the title.

> What we are trying to achieve in this PR is to get the block to early beta WPCOM users, so we can get these comments from them. We are aware of all issues regarding copy, phrasing, etc. But we are pursuing a select group of users to test it first. So we are 100% aware of the issues you are pointing out. However, instead of giving it our best guess we'd like to iron out those kinks with the users.

This pattern seems to be something that we keep bumping into. I'd love to have a discussion around process. It seems like PR's might benefit from having a simple label of either #ship-fast-and-learn or #fully-baked-solution or something like that. Otherwise it's too easy to make assumptions and too hard to know what sort of feedback is being requested.

simison commented:
>This pattern seems to be something that we keep bumping into. I'd love to have a discussion around process. It seems like PR's might benefit from having a simple label of either #ship-fast-and-learn or #fully-baked-solution or something like that. Otherwise it's too easy to make assumptions and too hard to know what sort of feedback is being requested.

Why not just indicate type of feedback requested in PR description?

It's especially important to communicate it when you're receiving feedback from external teams. Within your team these expectations might be clear from get-go but the rest of us cannot know unless you explicitly communicate it. ;-)

It can also be helpful to iterate & merge more often in much smaller PRs behind the beta flag, as it typically avoids frustrations like these for all parties involved. Communicating that you're intending to loop back to address feedback later on (and then really follow up of course) allows you to merge faster, too. Rarely feedback is really meant as blocking unless explicitly said so. :-)

artpi commented:
> Why not just indicate type of feedback requested in PR description?

Good idea, that was missing from the description. Added a header 👍 

> It can also be helpful to iterate & merge more often in much smaller PRs behind the beta flag, as it typically avoids frustrations like these for all parties involved.

Fantastic. Let's do that!

> Communicating that you're intending to loop back to address feedback later on (and then really follow up of course) allows you to merge faster, too.

Yes, I am trying to do so,  but I'll try to be more explicit.

> Rarely feedback is really meant as blocking unless explicitly said so. :-)

Well, lack of approval is blocking and there is not much we can do about it. But we will double down on communicating that we will adress stuff later!

Thanks !

jeherve commented:
Happy to merge this once the blockers are out of the way, and we can iterate and address some of those other questions during the Beta testing phase.

Here is a quick list of the comments from my review above that I would consider blockers:

- [ ] PHP Notices when creating new products
- [ ] Remove file from `module-extras.php` https://github.com/Automattic/jetpack/pull/9802#discussion_r278229941
- [ ] Update file to match WordPress.com https://github.com/Automattic/jetpack/pull/9802#discussion_r278567082
- [ ] `Jetpack::is_active()` checks https://github.com/Automattic/jetpack/pull/9802#discussion_r278237267 https://github.com/Automattic/jetpack/pull/9802#discussion_r278269453
- [ ] Rebase to fix phpcs whitelist conflicts, and fixing the list at the same time https://github.com/Automattic/jetpack/pull/9802#discussion_r278238313
- [ ] Better error messages for folks using the block when not connected to WordPress.com https://github.com/Automattic/jetpack/pull/9802#discussion_r278564025

artpi committed:
Review feedback: remove unnecessary import

artpi committed:
Better feedback where my account is not connected

artpi committed:
Endpoints nly for wpcom/active jetpack

artpi committed:
remove unused function

artpi commented on _inc/lib/core-api/wpcom-endpoints/memberships.php:
```diff
@@ -0,0 +1,170 @@
+		);
+	}
+}
+wpcom_rest_api_v2_load_plugin( 'WPCOM_REST_API_V2_Endpoint_Memberships' );
```
Done!
https://github.com/Automattic/jetpack/pull/9802/commits/4478d5c628e8db443d2ff4a1be003618569eb403

artpi commented:
> If that can help, here is a quick list of the comments from my review above that I would consider blockers:

That is fantastic. Thank you for that! 💯 

> PHP Notices when creating new products

That seems to be a broader issue with Jetpack API. I made a bug report here: https://github.com/Automattic/jetpack/issues/12183

I implemented all the other issues :)

artpi committed:
Additional check if JP is active

artpi committed:
Pass Lang to subscribe form

artpi committed:
error catching

jeherve commented:
> That seems to be a broader issue with Jetpack API. I made a bug report here: #12183

👍 I replied there with a potential way to address that.

kraftbj commented:
With #12183, let's figure out today if setting `type` is enough to get this PR going and we can address hardening 12183 after today's freeze OR if we do need to do work in 12183 for 7.3, we can merge this now and make sure 12183 is fixed during the beta week.

kraftbj commented:
The API issue isn't restricted to this PR. @jeherve is resolving the conflict and we'll merge this now. The API can be addressed during beta.

jeherve committed:
PHPCS whitelist: add the whole new module folder to the pre-commit

jeherve reviewed (approved):
Let's merge this now and address the feedback in future PRs.

It's worth noting that I'm currently having issues inserting new amounts. I get the following JavaScript error when I try to submit a new amount:

```
Uncaught (in promise) TypeError: Cannot read property 'id' of null
    at eval (edit.jsx:194)
```

This matches the following error on WordPress.com:

```
Fatal error: Uncaught Error: Call to undefined method WP_Error::to_array() in [...]jetpack-endpoints/memberships.php:105
```

kraftbj commented:
Merged, but let's definitely address the API this week (Crew) and this new error asap (@artpi).

BurlesonBrad commented:
I'm not wanting to go down the coding rabbit hole any more, but here's an innertubes show for ya to see and watch what happens:
http://youtu.be/SNxLOaHWAt8?hd=1

First, It says I'm **NOT** connected to the Pack-O-Jets.  Which, _**I am!!**_  I've paid for the Premium Jets:
![image](https://user-images.githubusercontent.com/6208757/57485604-bd0b0e00-7271-11e9-8d6a-51d0e167a5a5.png)

After the "_You ain't connected to the Packs de La Jets_" prompt, the "Connect to Stripe" button can't be clicked at all.

> Great, first it prompts me that my Packs don't get to connect with any Jets. And now they don't get to have any Stripes either.  Awesomesauce. 😭 

I turned on my Beta Blockers and loaded up the Gutenberries as well 👍 

![image](https://user-images.githubusercontent.com/6208757/57485920-8b467700-7272-11e9-9b68-b2c3c784ded9.png)

> Well. That didn't help either 😢 

So, I decided to retry the Stripes, the Memberships, The Guttenberries, and the PacksOfJets again at 00:36 & still nothing https://youtu.be/SNxLOaHWAt8?t=36

jeherve commented:
@BurlesonBrad That error message should be displayed when the user you are logged in with is not linked to any WordPress.com account.

When you originally connected your site to WordPress.com, you did so while logged in as an admin on your site. If another admin wants to use the Memberships block, they'll need to link their own account to WordPress.com first. Does that make sense, and apply to you here, or do you see that error message while logged in with the admin account you used when you first connected your site to WordPress.com?

If the error appears even though the account you are currently logged in with is linked to WordPress.com, could you contact us [via this form](https://jetpack.com/contact-support/?rel=support) so we can run more tests with you and get this fixed?

Thank you!

BurlesonBrad commented:
Yessir.  In that quick vid and the screenshot as well, both are logged in and connected to the same WP.com account that I've had since dinosaurs roamed the planet.
